File: src/upc-util.test.js
//
// Copyright (c) DITUS INC. All rights reserved. See LICENSE file in the project
// root for details.
//
import UpcUtil from './upc-util';

describe('UpcUtil', () => {
  describe('calculateChecksum', () => {
    it('returns null if the UPC is null, undefined, or an empty string.', () => {
      expect(UpcUtil.calculateChecksum(null)).toBe(null);
      expect(UpcUtil.calculateChecksum(undefined)).toBe(null);
      expect(UpcUtil.calculateChecksum('')).toBe(null);
      expect(UpcUtil.calculateChecksum(' ')).toBe(null);
    });

    it('returns the expected checksum for EAN-13 codes.', () => {
      expect(UpcUtil.calculateChecksum('9501101530003')).toBe(3);
    });

    it('returns the expected checksum for UPC-A codes.', () => {
      expect(UpcUtil.calculateChecksum('614141000036')).toBe(6);
    });

    it('returns the expected checksum for EAN-8 codes.', () => {
      expect(UpcUtil.calculateChecksum('95050003')).toBe(3);
    });

    it('returns the expected checksum for UPC-E codes.', () => {
      expect(UpcUtil.calculateChecksum('06107420')).toBe(0);
    });
  });

  describe('generateTestUpc', () => {
    it('returns a UPC where the first six digits are 099999.', () => {
      expect(UpcUtil.generateTestUpc().substring(0, 6)).toBe('099999');
    });

    it('returns a UPC where the length is 12 digits.', () => {
      expect(UpcUtil.generateTestUpc().length).toBe(12);
    });

    it('returns a UPC that is different with each call.', () => {
      for (let i = 1; i <= 100; i += 1) {
        const first = UpcUtil.generateTestUpc();
        const second = UpcUtil.generateTestUpc();
        expect(first).not.toBe(second);
      }
    });

    it('returns a UPC that is a valid UPC.', () => {
      for (let i = 1; i <= 100; i += 1) {
        expect(UpcUtil.isValid(UpcUtil.generateTestUpc())).toBeTruthy();
      }
    });
  });

  describe('isValid', () => {
    it('returns false if the UPC is null, undefined, or an empty string.', () => {
      expect(UpcUtil.isValid(null)).toBeFalsy();
      expect(UpcUtil.isValid(undefined)).toBeFalsy();
      expect(UpcUtil.isValid('')).toBeFalsy();
      expect(UpcUtil.isValid(' ')).toBeFalsy();
    });

    it('returns true for valid EAN-13 codes.', () => {
      expect(UpcUtil.isValid('9501101530003')).toBeTruthy();
    });

    it('returns true for valid UPC-A codes.', () => {
      expect(UpcUtil.isValid('614141000036')).toBeTruthy();
      expect(UpcUtil.isValid('099999109955')).toBeTruthy();
    });

    it('returns true for valid EAN-8 codes.', () => {
      expect(UpcUtil.isValid('95050003')).toBeTruthy();
    });

    it('returns true for valid UPC-E codes.', () => {
      expect(UpcUtil.isValid('06107420')).toBeTruthy();
    });
  });
});

File: DotNet/Ditus.Upc.Test/UpcUtilTest.cs
// Copyright (c) DITUS INC. All rights reserved. See LICENSE file in the project
// root for details.
namespace Ditus.Upc.Test;

using Microsoft.VisualStudio.TestTools.UnitTesting;

/// <summary>
/// Contains methods for testing the <see cref="UpcUtil"/> class.
/// </summary>
[TestClass]
public class UpcUtilTest
{
  /// <summary>
  /// Tests <see cref="UpcUtil.CalculateChecksum(string)"/>.
  /// </summary>
  [TestMethod]
  public void CalculateChecksumTest()
  {
    Assert.IsNull(UpcUtil.CalculateChecksum(null), "Returns null when the UPC is null.");
    Assert.IsNull(UpcUtil.CalculateChecksum(string.Empty), "Returns null when the UPC is an empty string.");
    Assert.IsNull(UpcUtil.CalculateChecksum(" "), "Returns null when the UPC is whitespace.");
    Assert.AreEqual(3, UpcUtil.CalculateChecksum("9501101530003"), "Returns the expected checksum for EAN-13 codes.");
    Assert.AreEqual(6, UpcUtil.CalculateChecksum("614141000036"), "Returns the expected checksum for UPC-A codes.");
    Assert.AreEqual(3, UpcUtil.CalculateChecksum("95050003"), "Returns the expected checksum for EAN-8 codes.");
    Assert.AreEqual(0, UpcUtil.CalculateChecksum("06107420"), "Returns the expected checksum for UPC-E codes.");
  }

  /// <summary>
  /// Tests <see cref="UpcUtil.IsValid(string)"/>.
  /// </summary>
  [TestMethod]
  public void IsValidTest()
  {
    Assert.IsFalse(UpcUtil.IsValid(null), "Returns false when the UPC is null.");
    Assert.IsFalse(UpcUtil.IsValid(string.Empty), "Returns false when the UPC is an empty string.");
    Assert.IsFalse(UpcUtil.IsValid(" "), "Returns false when the UPC is whitespace.");
    Assert.IsTrue(UpcUtil.IsValid("9501101530003"), "Returns true when a valid EAN-13 code is specified.");
    Assert.IsTrue(UpcUtil.IsValid("614141000036"), "Returns true when a valid UPC-A code is specified.");
    Assert.IsTrue(UpcUtil.IsValid("95050003"), "Returns true when a valid EAN-8 code is specified");
    Assert.IsTrue(UpcUtil.IsValid("06107420"), "Returns true when a valid UPC-E code is specified");
  }
}
